### PR TITLE
Terminate Policy with End Date

### PR DESCRIPTION
**Pull request by rshade**
### Description
Terminates instances after the expiration of an end date tag. 

### Issues Resolved

[List any existing issues this PR resolves]

### Contribution Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable
- [ ] New functionality has been documented in CHANGELOG.MD

### CODE REVIEW

File: cost/terminate_policy/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+***instance:terminate Tag Example***
+`instance:terminate=2019-05-20T13:48:21Z`
+
+
+### Input Parameters
+
+This policy has the following input parameters required when launching the policy.
+
+- *Email addresses* - AEmail addresses of the recipients you wish to notify
```
Comment (cdwilhelm) [inline cost/terminate_policy/README.md]: Typeo on AEmail, but should be An email

File: cost/terminate_policy/server_terminate.pt [hunk 2]
```diff
@@ -0,0 +1,188 @@
+name "Terminate Instances with End Date"
+rs_pt_ver 20180301
+type "policy"
+short_description "This Policy Template is used to automatically resize AWS Instances based on user-defined standards. \n See the [README](https://github.com/rightscale/policy_templates/tree/master/cost/aws/aws_superseded_instance) and [docs.rightscale.com/policies](http://docs.rightscale.com/policies/) to learn more."
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: Update the short description for THIS policy

File: cost/terminate_policy/server_terminate.pt [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+###############################################################################
+# Parameters
+###############################################################################
+
+parameter "param_email" do
+  type "list"
+  label "Email addresses"
+  description "Email addresses of the recipients you wish to notify"
+end
+
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: I believe we decided all policies should have an exclude tag option.    Any reason why this shouldn't have an exclude tag?

File: cost/terminate_policy/server_terminate.pt [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+# Authentication
+###############################################################################
+
+auth "auth_rs", type: "rightscale"
+
+###############################################################################
+# Resources
+###############################################################################
+
+resources "clouds", type: "rs_cm.clouds"
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: The readme lists three clouds supported.  maybe the readme should include some thing like All rightscale supported clouds or do a filter here to limit to those in the readme.

File: cost/terminate_policy/server_terminate.pt [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+resources "instances", type: "rs_cm.instances" do
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: Only terminate operational, what about running or stopped/provisioned/stranded.  probably need to update the readme to indicate what instance states are valid for terminate

File: cost/terminate_policy/server_terminate.pt [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+  iterate @clouds  
+  cloud_href href(iter_item)
+  tags 'instance:terminate=*'
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: This is a rightscale tag, and not everyone is using CMP.  I'm thinking maybe the tag should be configurable, like the downsize policy.  what about providing a parameter to filter by ec2:instance, gce:instance and azure:instance or something like that?

File: cost/terminate_policy/server_terminate.pt [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+script "js_instances_with_timestamp", type: "javascript" do
+  parameters "ds_instances"
+  result "results"
+  code <<-EOF
+    var results = []
+    var date = new Date()
+    var iso_date = date
+    for ( i = 0; i < ds_instances.length; i++ ) {
+      var instance = ds_instances[i]
+      if ( !instance.locked ){
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: readme should be updated to indicate locked instances are ignored

File: cost/terminate_policy/server_terminate.pt [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        var instance_timestamp = instance.tags.split("=")[1];
+        var instance_terminate_date = new Date(instance_timestamp);
+        console.log(typeof(instance_terminate_date))
+        var instance_terminate_date_iso = instance_terminate_date
+        console.log(instance_terminate_date+"-"+iso_date)
+        var should_terminate = false
+        if (  instance_terminate_date <= iso_date ) {
+          console.log("true")
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: remove debuging comments

File: cost/terminate_policy/server_terminate.pt [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        
+        results.push({
+          resource_uid: instance.resource_uid,
+          name: instance.name,
+          href: instance.href
+          cloud_href: instance.cloud_href,
+          instance_type_href: instance.instance_type_href,
+          locked: instance.locked,
+          state: instance.state,
+          tags: instance.tags,
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: do you really want to show all tags, or just the terminate tag value, like terminate date?

File: cost/terminate_policy/server_terminate.pt [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+policy "pol_instances" do
+  validate_each $ds_instances_with_timestamp do
+    summary_template "{{ rs_project_name }} (Account ID: {{ rs_project_id }}): {{ len data }} Terminable Servers"
+    detail_template <<-EOS
+# Unattached IP Addresses
+| Name | Terminate_Tag | Current Time |
+| ---- | ------------- | ------------ |
+{{ range data -}}
+| {{ .name }} | {{.tags}} | {{.run_time}} |
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: for terminate tag, changed to terminate date and show the terminate tag value.

Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: for terminate tag, changed to terminate date and show the terminate tag value.

File: cost/terminate_policy/server_terminate.pt [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    end
+  end
+end
+
+define sys_log($subject, $detail) do
+  if $$debug
+    rs_cm.audit_entries.create(
+      notify: "None",
+      audit_entry: {
+        auditee_href: @@account,
```
Comment (cdwilhelm) [inline cost/terminate_policy/server_terminate.pt]: the subject should include the policy name so it's easy to find in the audit entries.

File: cost/terminate_policy/server_terminate.pt (hunk 3 above)
Comment (rshade) [reply cost/terminate_policy/server_terminate.pt]: Do we need to exclude items that have an implicit include?

File: cost/terminate_policy/server_terminate.pt (hunk 5 above)
Comment (rshade) [reply cost/terminate_policy/server_terminate.pt]: I am removing the filter

File: cost/terminate_policy/server_terminate.pt (hunk 9 above)
Comment (rshade) [reply cost/terminate_policy/server_terminate.pt]: I filter to just that tag in the data source.

File: cost/terminate_policy/instance_terminate.pt [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+  code <<-EOF
+    var results = []
+    var date = new Date()
+    var iso_date = date
+    for ( i = 0; i < ds_instances.length; i++ ) {
+      var instance = ds_instances[i]
+      tags = instance.tags
+      for (n = 0; n < tags.length; n++){
+        tag = tags[n]
+        console.log(tag)
```
Comment (cdwilhelm) [inline cost/terminate_policy/instance_terminate.pt]: remove debug logging

File: cost/terminate_policy/instance_terminate.pt [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+        tag_predicate = tag.split('=')[0];
+        console.log(tag_predicate)
```
Comment (cdwilhelm) [inline cost/terminate_policy/instance_terminate.pt]: remove

**Pull request merged**